### ISSUE TITLE
Optimize sql plans caching with varchar\nvarchar size

### ISSUE THREAD

**Issue by arwyl** [linq2db/linq2db#989]
I was wondering, is there a way to set the size of all nvarchar and varchar parameters to max?
I managed to do this with DataConnection.OnTrace event, but maybe there is a better way.

```cs
Linq2DBUtils.SwitchTraceOn(onBeforeExecute: x =>
			{
				foreach (var par in x.Command.Parameters)
				{
					var sqlP = par as SqlParameter;
					if (sqlP != null && sqlP.SqlDbType == System.Data.SqlDbType.NVarChar)
					{
						sqlP.Size = -1;
					}

					if (sqlP != null && sqlP.SqlDbType == System.Data.SqlDbType.VarChar)
					{
						sqlP.Size = -1;
					}
				}
			});

			using (var db = _lifeTimeScope.Resolve<TestDataConnection>())
			{
				int id = 123;
				string name = "someTest";
				bool isActive = true;
				db.Update<User>(x => x.Id == id, x => new User
				{
					Name = name,
					IsActive = isActive
				});
                        }
```

this code leads to following query in sql profiler:
```sql
(@name nvarchar(max) ,@isActive bit,@id int)UPDATE
	[t1]
SET
	[t1].[Name] = @name,
	[t1].[IsActive] = @isa
FROM
	[Users] [t1]
WHERE
	[t1].[Id] = @id
```
So it's plan will be cached and used for following updates.
Without setting the Size to -1 the query will be:
```sql
(@name nvarchar(8) ,@isActive bit,@id int)UPDATE
	[t1]
SET
	[t1].[Name] = @name,
	[t1].[IsActive] = @isa
FROM
	[Users] [t1]
WHERE
	[t1].[Id] = @id
```
and it's plan will be used only for updates where name length is only 8

**Comment by sdanyliv** [comment]
@arwyl, which DataProvider do you use?

**Comment by arwyl** [comment]
@sdanyliv SqlServer.2012

**Comment by sdanyliv** [comment]
@arwyl, i have prepared fix. How to ensure that plan is properly cached?

**Comment by sdanyliv** [comment]
Nevermind, i've found how to check cached plans
```sql
SELECT plan_handle, query_plan, objtype   
FROM sys.dm_exec_cached_plans   
CROSS APPLY sys.dm_exec_query_plan(plan_handle)   
where objtype = 'Prepared'

--DBCC FREEPROCCACHE
```
Should we take in account other db types also? Which of them?

**Comment by arwyl** [comment]
If the resulting sql is equal, then plan is cached =)

In this query you can see the resulting sql in [Parent Query] field
```
select top 50 
    [max worker time (s)] = cast (qs.max_worker_time * 1. / 1e6 as decimal(18,2)),
    [total elapsed time (s)] = cast(qs.total_elapsed_time * 1. / 1e6 as decimal(18,2)),
    qs.total_rows, 
    [max elapsed time (s)] = cast (qs.max_elapsed_time * 1. / 1e6 as decimal(18,2)),
    [Average CPU used (sec)] = cast(qs.total_worker_time * 1. / qs.execution_count / 1e6 as decimal(18,3)),
    [Total CPU used (sec)] = cast (qs.total_worker_time * 1. / 1e6 as decimal(18,2)),
    [Execution count] = qs.execution_count,
    [Individual Query] = substring (qt.text,qs.statement_start_offset/2+1, (case when qs.statement_end_offset = -1 then len(convert(nvarchar(max), qt.text)) * 2 else qs.statement_end_offset end - qs.statement_start_offset)/2+1),
    [Parent Query] = qt.text,
    DatabaseName = db_name(qt.dbid),
    [exec/hr] = cast(3600. * qs.execution_count / nullif(datediff(ss,qs.creation_time,getdate()),0) as decimal(18,2)),
    qp.query_plan,
	qs.last_execution_time
from  sys.dm_exec_query_stats qs
    outer apply sys.dm_exec_sql_text(qs.sql_handle) as qt
    outer apply sys.dm_exec_query_plan(qs.plan_handle) qp
order by 
    qs.last_execution_time desc;
```

I've also adapted my test not to depend on my other code. Maybe it helps
```
using System;
using System.Collections.Generic;
using System.Linq;
using NUnit.Framework;

namespace Tests.SQL
{
	[TestFixture]
	public class SqlParametersTest
	{
		private List<string> GetLastQueriesFromSql(int count, string containsText)
		{
			using (var db = GetConnection())
			{
						var queries = db.Query<string>(@"select top " + count + @"
			--qs.*,
			--    [max worker time (s)] = cast (qs.max_worker_time * 1. / 1e6 as decimal(18,2)),
			--    [total elapsed time (s)] = cast(qs.total_elapsed_time * 1. / 1e6 as decimal(18,2)),
			--    qs.total_rows, 
			--    [max elapsed time (s)] = cast (qs.max_elapsed_time * 1. / 1e6 as decimal(18,2)),
			--    [Average CPU used (sec)] = cast(qs.total_worker_time * 1. / qs.execution_count / 1e6 as decimal(18,3)),
			--    [Total CPU used (sec)] = cast (qs.total_worker_time * 1. / 1e6 as decimal(18,2)),
			--    [Execution count] = qs.execution_count,
			--    [Individual Query] = substring (qt.text,qs.statement_start_offset/2+1, (case when qs.statement_end_offset = -1 then len(convert(nvarchar(max), qt.text)) * 2 else qs.statement_end_offset end - qs.statement_start_offset)/2+1),
				[Parent Query] = qt.text
			--    DatabaseName = db_name(qt.dbid),
			--    [exec/hr] = cast(3600. * qs.execution_count / nullif(datediff(ss,qs.creation_time,getdate()),0) as decimal(18,2)),
			--    qp.query_plan
			from  sys.dm_exec_query_stats qs
				outer apply sys.dm_exec_sql_text(qs.sql_handle) as qt
				outer apply sys.dm_exec_query_plan(qs.plan_handle) qp
			where qt.text like '%' + @ContainsText + '%'
			order by 
				qs.last_execution_time desc;", Utils.QuickParams("ContainsText", containsText));
						return queries;
			}
		}

		IYourDataConnection GetConnection()
		{
			// implementation
		}

		[Test]
		public void SqlParameters_Varchar_NVarchar_Size_Is_Always_Max_Linq2Db()
		{
			using (var db = GetConnection())
			{
				// call query several times with different parameter length
				for (int i = 0; i < 10; i++)
				{
					string someName = "testtest" + (10 * i);
					// hardcode a guid in query to find it in sys.dm_exec_sql_text
					var users = db.Users.Where(x => x.Name == someName && x.Guid.ToString() == "4b18d00b-0fe4-436a-a2e5-7d4fa8231aa2").Select(x => x.Id).ToList();
				}
			}

			var queries = GetLastQueriesFromSql(1, "4b18d00b-0fe4-436a-a2e5-7d4fa8231aa2");

			Assert.NotNull(queries);
			Assert.IsNotEmpty(queries);

			// nvarchar(max), not nvarchar(some number)
			Assert.IsTrue(queries[0].StartsWith("(@someName nvarchar(max)"), queries[0]);
		}
	}
}

```

**Comment by sdanyliv** [comment]
No problem, thanks for detailed explanations.

**Comment by arwyl** [comment]
@sdanyliv am I correct, that for release 1.10.1 there are no ways to do that except for using DataConnection.OnTrace event?

**Comment by sdanyliv** [comment]
Yes. You have found the only way.
Or you can create your own data provider.

**Comment by arwyl** [comment]
@sdanyliv our tests found a problem with nvarchar(max) and containstable https://docs.microsoft.com/en-Us/sql/t-sql/queries/contains-transact-sql

So it seems to set nvarchar length to 4000 and varchar length to 8000 respectively, but not to max

**Comment by sdanyliv** [comment]
Which prolem? It’s silently truncated to 4000?

**Comment by arwyl** [comment]
@sdanyliv no, i mean that query like 
```
declare @str nvarchar(max)
select * from ContainsTable(SomeTable, SomeColumnWithFullTextIndex, @str)
```
will cause an error

but query
```
declare @str nvarchar(4000)
select * from ContainsTable(SomeTable, SomeColumnWithFullTextIndex, @str)
```
won't

so parameter.Size = -1 should be changed for Size = 4000 for nvarchar and Size = 8000 for varchar

**Comment by arwyl** [comment]
@sdanyliv one more thing:
```
if(parameter.DbType == SqlDbType.NVarChar)
{
// also add null check
    parameter.Size = parameter.Value.ToString().Length > 4000 ? -1 : 4000;
}
if(parameter.DbType == SqlDbType.VarChar)
{
// also add null check
    parameter.Size = parameter.Value.ToString().Length > 8000 ? -1 : 8000;
}
```

because there really can be field with max size..

**Comment by sdanyliv** [comment]
Added fix for that in PR #990
